src/main/java/com/ita/if103java/ims/service/impl/WarehouseServiceImpl.java

```diff
@@ -66,20 +94,39 @@ public WarehouseDto findWarehouseById(Long id) {
     }
 
     @Override
-    public List<WarehouseDto> findWarehousesByTopLevelId(Long topLevelId) {
+    public List<WarehouseDto> findWarehousesByTopLevelId(Long topLevelId, UserDetailsImpl user) {
         return warehouseDtoMapper.toDtoList(warehouseDao.findChildrenByTopWarehouseID(topLevelId));
     }
 
     @Override
-    public WarehouseDto update(WarehouseDto warehouseDto) {
+    public WarehouseDto update(WarehouseDto warehouseDto, UserDetailsImpl user) {
         Warehouse updatedWarehouse = warehouseDtoMapper.toEntity(warehouseDto);
         Warehouse dBWarehouse = warehouseDao.findById(updatedWarehouse.getId());
         updatedWarehouse.setActive(dBWarehouse.isActive());
+        EventName eventName = EventName.WAREHOUSE_EDITED;
+        createEvent(user, updatedWarehouse, eventName);
         return warehouseDtoMapper.toDto(warehouseDao.update(updatedWarehouse));
     }
 
     @Override
-    public boolean softDelete(Long id) {
-        return warehouseDao.softDelete(id);
+    public boolean softDelete(Long id, UserDetailsImpl user) {
+        Boolean isDelete = warehouseDao.softDelete(id) ;
+        if (isDelete){
+        EventName eventName = EventName.WAREHOUSE_REMOVED;
+        Warehouse warehouse = warehouseDao.findById(id);
+        createEvent(user, warehouse, eventName);
```

Comment:
Can be simplified
createEvent(user, warehouse, EventName.WAREHOUSE_REMOVED);

Please add event message "Removed Warehouse '{name}' with level '{level}'"
For example:
Removed Warehouse 'Storage' with level '0'